src/Microsoft.VisualStudio.ProjectSystem.Managed/ProjectSystem/LanguageServices/CommandLineArgumentsProvider.cs

```diff
@@ -0,0 +1,95 @@
+﻿// Licensed to the .NET Foundation under one or more agreements. The .NET Foundation licenses this file to you under the MIT license. See the LICENSE.md file in the project root for more information.
+
+using System;
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.ComponentModel.Composition;
+using System.Threading.Tasks.Dataflow;
+using Microsoft.VisualStudio.Composition;
+using Microsoft.VisualStudio.ProjectSystem.Build;
+
+namespace Microsoft.VisualStudio.ProjectSystem.LanguageServices
+{
+    [Export(typeof(ICommandLineArgumentsProvider))]
+    [AppliesTo(ProjectCapability.DotNet)]
+    internal sealed class CommandLineArgumentsProvider : ChainedProjectValueDataSourceBase<CommandLineArgumentsSnapshot>, ICommandLineArgumentsProvider
+    {
+        private readonly IProjectBuildSnapshotService _projectBuildSnapshotService;
+
+        [ImportingConstructor]
+        public CommandLineArgumentsProvider(
+            ConfiguredProject project,
+            IProjectBuildSnapshotService projectBuildSnapshotService)
+            : base(project, synchronousDisposal: false, registerDataSource: false)
+        {
+            _projectBuildSnapshotService = projectBuildSnapshotService;
+        }
+
+        protected override IDisposable? LinkExternalInput(ITargetBlock<IProjectVersionedValue<CommandLineArgumentsSnapshot>> targetBlock)
+        {
+            CommandLineArgumentsSnapshot? lastSnapshot = null;
+
+            // Transform the changes from build snapshot to command line arguments.
+            // We support skipping inputs/output values if multiple exist (WithNoDelta), as only the most recent values are deemed useful.
+            DisposableValue<ISourceBlock<IProjectVersionedValue<CommandLineArgumentsSnapshot>>> transformBlock
+                = _projectBuildSnapshotService.SourceBlock.TransformWithNoDelta(update => update.Derive(snapshot => Transform(snapshot)));
+
+            // Set the link up so that we publish changes to target block
+            transformBlock.Value.LinkTo(targetBlock, DataflowOption.PropagateCompletion);
+
+            // Join the source block, so if it needs to switch to UI thread to complete 
+            // and someone is blocked on us on the same thread, the call proceeds
+            JoinUpstreamDataSources(_projectBuildSnapshotService);
+
+            return transformBlock;
+
+            CommandLineArgumentsSnapshot Transform(IProjectBuildSnapshot buildSnapshot)
```

Comment:
While I understand the usefulness of writing code like this as local functions, the clarity of scoping becomes messy after you start nesting local functions within local functions (like `IsUnchanged`) below. Maybe I'm just an old-head, but having logic like this as *minimally scoped as possible* seems like a good way to keep code intent clear. For example, if this code does not use class instance variable, it should be `static`. If all information can simply be passed into it instead of containing the scope of the outer method, it can just become its own `static method`. In this local function, the only value required is already being passed in as a parameter, `buildSnapshot`. The `lastSnapshot` variable is accessed from the method scope, but (as far as I can tell) the method itself doesn't use `lastSnapshot` at all. The only benefit is that `lastSnapshot` will exist for the duration of the method scope, but I can't tell if that is intended by the code available here. Similarly, `IsUnchanged` requires an `ImmutableArray<string>` (for snapshot arguments) and a target outputs (from the buildSnapshot) which could be passed into it, thus not needing any scopes above it, which makes it a `private static method` in my mind. This increases the ability (in my option) to easily understand that the code in that block only needs to be read for the complete picture; I don't need to look elsewhere. It is self-contained. The only reason I'm not a fan of this local functions pattern is I had past experience writing in ColdFusion, where the entire language has this scoping issue, but much, much worse. There were some bits of code I was refactoring that was 10 scopes deep, and you'd have to look through a few thousand line file to find where a certain variable was created and when it was actually set with useful information. Here, this is obviously not as complex as that, but I kinda see this as an anti-pattern in any language to unnecessarily scope code that can be totally independent. Here, the fact it is local is primarily used to associate that block with this method directly, which I think is a good thing, but not at the price of pulling in the method's scope. Additionally, this style causes more and more intending which I'm also not a fan of. Generally, the deeper the indentation, the more smelly the code is. That isn't always the case, but generally smaller, more encapsulated code more left justified. Lastly, the testability of these local functions is basically non-existent as you can't access them outside of the method. At least with static methods in the class, you can `InternalsVisisbleTo` so unit test projects can access them.  Personally, I think I'd only use local functions if that are very small and have multiple callsites in the method. Here, each of these local functions are only called once by their outer scope. That's just my 2-cents.

Reply:
There are trade-offs and style preferences. I like this kind of nesting as it limits the number of things I have to think about while reading the code. In general I find knowing where a bit of code can be called from to be the most useful thing. In this class, it's clear that all logic happens in response to a call to `LinkExternalInput`.

If each were moved into separate top-level functions it would become harder to work out how they relate to each other. This can become especially problematic in larger classes. Nesting functions keeps scopes tight and clear.